### ISSUE TITLE
Implement semi-supervised clustering with HDBSCAN

### ISSUE THREAD

**Issue by azizkayumov** [petabi/petal-clustering#87]
HDBSCAN's [semi-supervised extension](https://link.springer.com/article/10.1007/s10618-019-00651-1) would be a great addition to the library. For example, given a partially-labeled data (e.g. 10-20% of points are labeled by the user), we would like to obtain clustering results that best fits the points of the same class into the same clusters as well as discover novel clusters based on density.

Currently, there are two implementations of semi-supervised clustering using HDBSCAN, in [Java](https://github.com/jadsoncastro/UnifiedView) and [Python](https://github.com/TutteInstitute/fast_hdbscan). The Python version is quite recent, but its minimum spanning tree looks to be approximate. So I was thinking of implementing the framework for `HDbscan`, based on either of the following options. 

### Option A.
[`Predict`](https://github.com/petabi/petal-clustering/blob/7e077bd2c9d1e5bc539468f185b2f68b51f3ec44/src/lib.rs#L18) trait can be changed to the following and then implemented by HDBSCAN:
```
/// An interface to apply a trained model.
pub trait Predict<I, O>
where
    I: ?Sized,
{
    fn predict(&mut self, input: &I, partial_labels: O) -> O;
}
```
But this change may not follow the same structure of the Python [implementation](https://github.com/TutteInstitute/fast_hdbscan/blob/519fb3f3feb2f5488c830ea177a4ed5887cd5252/fast_hdbscan/hdbscan.py#L318), where they added `partial_labels` as optional parameter to the `fit` function itself.

### Option B.
Change `Fit` trait by adding `fit_predict` function as in scikit-learn's [DBSCAN](https://github.com/scikit-learn/scikit-learn/blob/2c0cdd479a2d229e88afbb97cce91eef3d1b07c9/sklearn/cluster/_dbscan.py#L448):
```
/// An interface to train a model.
pub trait Fit<I, O>
where
    I: ?Sized,
{
    fn fit(&mut self, input: &I) -> O;
    fn fit_predict(&mut self, input: &I, partial_labels: &O) -> O;
}
```
This will force unnecessary changes in `Optics` or `Dbscan` structs to implement the `fit_predict` function.

### Option C:
Add a new trait called `FitPredict` as in scikit-learn's [HDBSCAN](https://github.com/scikit-learn/scikit-learn/blob/2c0cdd479a2d229e88afbb97cce91eef3d1b07c9/sklearn/cluster/_hdbscan/hdbscan.py#L882):
```
/// An interface to train and predict: 
pub trait FitPredict<I, O>
where
    I: ?Sized,
{
    fn fit_predict(&mut self, input: &I, partial_labels: &O) -> O;
}
```
This will not introduce unnecessary forced changes in other structs and does seem semantically correct to me: we are fitting the unknown data and predicting their clusters based on partial labels.

Please let me know if you would be interested in including semi-supervised clustering for HDBSCAN, and which of the aforementioned options would be the best way to go.

**Comment by msk** [comment]
Thanks for the proposal! I agree that this would be an excellent addition to petal-clustering.

I'm particularly interested in this capability since I have a use case involving streaming data that could benefit greatly. In my application, I need to perform incremental clustering (processing the first batch, waiting for next batches, clustering and merging, etc.) since processing the entire dataset at once isn't feasible.

Regarding the interface design, have you considered simply adding an `Option` parameter to the existing `fit` function? Since clustering generally only has the "fit" stage (unlike other ML models with separate train/predict phases), having labels in `fit` would align well with the model fitting concept and maintain consistency with the Python implementation you referenced.

If there's a specific reason that approach won't work, I think Option A in your proposal makes the most sense for our codebase.

**Comment by azizkayumov** [comment]
Thanks for the suggestion, incremental clustering sounds really interesting when mixed with semi-supervision. Being able to use partial labels from previous batch (e.g. 10% points from prev. clustered results) to perform clustering on the next batch would save us from clustering the whole dataset at once!

**Comment by azizkayumov** [comment]
To check if our implementation matches with Python [variant](https://github.com/TutteInstitute/fast_hdbscan), I tested both algorithms on [Digits dataset](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_digits.html). 
Here is a summary of ARI scores:
```
ARI: 0.4859 (HDBSCAN vs. Ground truth)
ARI: 0.5368 (PyHDBSCAN-BC (10%) vs. Ground truth)
ARI: 0.5357 (RustHDBSCAN-BC (10%) vs. Ground truth)
ARI: 0.9816 (PyHDBSCAN-BC (10%) vs. RustHDBSCAN-BC (10%))
```
Both Rust and Python versions of the semi-supervised algorithms are improving unsupervised results a bit with 10% partially-labelled data, I think PCA can help in this case to lower the dimensionality and run clustering again.

**Comment by azizkayumov** [comment]
@msk Please let me know if the changes look good to you. The build check related to ubuntu kept failing somehow, I am guessing it is requesting to increase `rustc` version to at least `1.81`:
```
error: package `half v2.5.0` cannot be built because it requires rustc 1.81 or newer, while the currently active rustc version is 1.74.0
Either upgrade to rustc 1.81 or newer, or use
cargo update half@2.5.0 --precise ver
where `ver` is the latest version of `half` supporting rustc 1.74.0
```
Would it be appropriate if I change it in this PR? I will revert it back to `1.74` if it does not align well with other dependencies.
